### PR TITLE
GBRForest in TMVAEvaluator + SoftLepton taggers using GBRForest

### PR DESCRIPTION
**Pull request by ferencek**
This PR add GBRForest forest use in the TMVAEvaluator and switches SoftLepton taggers to using the GBRForest.

This is only the first step in switching the SoftLepton taggers to GBRForest. Next, we need to figure out how to add the GBRForest payloads to the CondDB and then we can complete the switch.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ferencek** [comment]
@Dr15Jones, now that soft lepton taggers were switched to `GBRForest`, they should be thread-safe. Can you please double-check that that is indeed the case.  Thanks.

Note, however, that we still keep the possibility of using TMVA, which is not thread-safe. So I was wondering whether it would be possible to move the whole `std::mutex` locking business from the `ElectronTagger` and `MuonTagger` code into the `TMVAEvaluator` code and enable locking in the `TMVAEvaluator::evaluate()` method only when `GBRForest` is not used.

**Comment by ferencek** [comment]
@Dr15Jones, I don't want to be a pest ;) but your input here would be greatly appreciated. Thank you.

**Comment by Dr15Jones** [comment]
@ferencek I'm sorry I haven't been able to respond. Unfortunately, there is an activity here at FNAL which is taking my full attention (even beyond full) all this week. I'm not sure I will be able to put any quality time into thinking about your question until this weekend.

**Comment by ferencek** [comment]
@Dr15Jones, no problem. Take your time, we are not in a particular hurry. I just wanted to make sure this was on your radar ;) Thanks.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10543	a3ee421
- code changes look OK
- jenkins tests pass and comparisons show no difference
- memory test [done in CMSSW_7_6_X_2015-08-06-2300 /test area sign576/ ] using PromptReco config on DoubleMuon-Run2015B-251721-44C77595 file shows reduction of 8 MB (traced with igprof to the change in TMVAEvaluator::initialize, as expected)

There is a static analyzer issue from this PR
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10543/6718/llvm-analysis/report-1b1c99.html#EndPath
, but I couldn't understand the logic of the analyzer that would actually lead to a problem

**Comment by ferencek** [comment]
@slava77, thanks for the summary. Let's then wait for the final comments from Chris Jones before we finalize this PR. Since there are no differences in the output, would you suggest a backport and to what release cycle(s) (I guess 75X should be fine but 74X maybe not). Thanks

**Comment by slava77** [comment]
Dinko,
if the same code change applies to 74X, please backport there as well.
It is a technical change reducing resource use, which is useful for 74X.

**Comment by ferencek** [comment]
OK, will do. Thanks

**Comment by ferencek** [comment]
Here is just a quick summary of the changes I had mind. Basically, the idea would be to change
```c++
    btag::LeptonSelector m_selector;
    mutable std::mutex m_mutex;
    [[cms::thread_guard("m_mutex")]] std::unique_ptr<TMVAEvaluator> mvaID;
```
in ` RecoBTag/SoftLepton/interface/MuonTagger.h ` to
```c++
    btag::LeptonSelector m_selector;
    std::unique_ptr<TMVAEvaluator> mvaID;
``
and in `CommonTools/Utils/interface/TMVAEvaluator.h` change
```c++
    std::unique_ptr<TMVA::Reader> mReader;

    std::map<std::string,float> mVariables;
    std::map<std::string,float> mSpectators;
```
to
```c++
    mutable std::mutex m_mutex;
    [[cms::thread_guard("m_mutex")]] std::unique_ptr<TMVA::Reader> mReader;

    [[cms::thread_guard("m_mutex")]] std::map<std::string,float> mVariables;
    [[cms::thread_guard("m_mutex")]] std::map<std::string,float> mSpectators;
```
Finally, the following line
```c++
std::lock_guard<std::mutex> lock(m_mutex);
```
would be removed from ` RecoBTag/SoftLepton/src/MuonTagger.cc` and
```c++
  if (!mUsingGBRForest)
    std::lock_guard<std::mutex> lock(m_mutex);
```
would be added to `TMVAEvaluator::evaluate()` in ` CommonTools/Utils/src/TMVAEvaluator.cc`. If and when @Dr15Jones confirms these changes would be sufficient to make the code thread-safe when `GBRForest` is used but still properly "locked" when TMVA is used, I will go ahead and make the changes.

**Comment by Dr15Jones** [comment]
Here is my suggestions

1) Make TMVAEvaluator a base class.
2) Have two classes inherit from it, TMVAEvaluatorWithReader and TMVAEvaluatorWithGBRForest
3) Make TMVAEvaluator::evaluate const. This will be needed to stop the static analyzer from complaining that MuonTagger is calling a non-const function.
4) Get rid of the initialize method in TMVAEvaluator and do the work in the constructor. Late initialization is extremely problematic plus you'd now want different ways to handle the initialization given the two different classes which would best be handled in the constructors.
4) Move the lock to be inside TMVAEvaluatorWithReader.

**Comment by ferencek** [comment]
@Dr15Jones, thanks for reviewing the code. Concerning making `TMVAEvaluator::evaluate` const, wouldn't that be a problem for `TMVAEvaluatorWithReader` where in the `evaluate()` call values of some of the data members need to be updated. Should those data members be made mutable?

**Comment by ferencek** [comment]
Another consideration is that there should also be the possibility of using GBRForest, which can be created on the fly from a TMVA Reader, in `TMVAEvaluatorWithReader`. This would essentially make the separation into `TMVAEvaluatorWithReader` and `TMVAEvaluatorWithGBRForest` somewhat artificial. Also, there might be cases where one wants to initialize the `TMVAEvaluator` in `beginRun()` so late initialization might be hard to avoid.

**Comment by ferencek** [comment]
Would the latest commit (https://github.com/cms-btv-pog/cmssw/commit/63590e356213ad3fddf6152acf20b28f9adca940) be sufficient to stop the static analyzer from complaining?

**Comment by slava77** [comment]
@cmsbuild please test
I suppose it will still complain about the mutables

**Comment by slava77** [comment]
The static analyzer issue is still there: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10543/7091/llvm-analysis/report-b9decf.html#EndPath

If I understand correctly, it wants you to check that ```mVariables.size() > 0```  (better define some ```const int nVariables = mVariables.size();``` 

As for the status of the PR, if I understand correctly, it is still not done:
- the latest commit implements mutex lock  which is not expected to work, if I understood Chris correctly (https://github.com/cms-sw/cmssw/pull/10543#commitcomment-12613249 )
- this should be followed by changes in MuonTagger and ElectronTagger (please clarify).

Due to the above, the signoff is on hold

**Comment by ferencek** [comment]
Hi Slava,

Apologies on not replying earlier. I was travelling today. Your observation was correct, there were still a few changes needed in the soft lepton taggers. These have been added now. I also tried to address the locking issue which I hope is now implemented correctly. I also made some updates to the float array but I'm not sure if the static analyzer will be happy about the updated version. You can restart tests and let's see how things look. Thanks.

**Comment by slava77** [comment]
@cmsbuild please test

File: CommonTools/Utils/src/TMVAEvaluator.cc [hunk 1]
```diff
@@ -76,17 +113,79 @@ float TMVAEvaluator::evaluate(const std::map<std::string,float> & inputs, const
...
+  return value;
+}
+
+
+float TMVAEvaluator::evaluateGBRForest(const std::map<std::string,float> & inputs) const
+{
+  // default value
+  float value = -99.;
+
+  float * vars = new float[mVariables.size()]; // allocate n floats
```
Comment (Dr15Jones) [inline CommonTools/Utils/src/TMVAEvaluator.cc]: Could we switch to ```std::unique_ptr<float[]>```?

**Comment by ferencek** [comment]
@Dr15Jones, I have implemented your suggestion about switching to `std::unique_ptr`.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by Dr15Jones** [comment]
This appears to be thread safe and the implementation is much cleaner than before.

**Comment by slava77** [comment]
... now on to figure out how to test it without das problems in the way

**Comment by ferencek** [comment]
Thanks, @Dr15Jones, for all the comments that brought the code to this stage.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10543	 0d350a8

- new code changes since https://github.com/cms-sw/cmssw/pull/10543	a3ee421 are OK
- jenkins tests went as far as was possible 
- tested locally using short++  matrix(https://github.com/slava77/cms-reco-tools/blob/master/runMatrix.sh)  and a prompt-reco setup for 200 events in DoubleMuon-Run2015B-251721 in CMSSW_7_6_0_pre3 /test area sign576/
    - in run2 prompt reco processing memory use decreased by about 7.5 MB, as in the previous test
```
1,850,824,034	6,594,785	main
2,225,544	 	MuonTagger::MuonTagger
5,936,004		ElectronTagger::ElectronTagger
==>
1,843,434,569	6,543,396	main
97,065	 	MuonTagger::initialize
362,432	 	ElectronTagger::initialize
```
    - monitored products didn't change, as expected

**Comment by slava77** [comment]
@cmsbuild please test
@monttj please take a look at this, AT signature is needed

**Comment by ferencek** [comment]
@gpetruc, @slava77, this PR cannot be backported to 74X before #10363 is merged. Should I just cherry-pick the commits from this PR and apply them to #10363?

**Comment by slava77** [comment]
Giovanni will be collecting 74X contributions in #10649, I see that 10363 is already there.
So, Giovanni should suggest what works better.
It looks like that PR (10649) is based on CMSSW_7_4_8_patch1

If your updates will be linear (no merges), you could probably update 10363.
Maybe you can start from CMSSW_7_4_8_patch1, make a "backport10543(pick a name)" branch, merge #10363 and then apply backport commits from this PR.
This will make a consistent PR with BTV features needed.

**Comment by ferencek** [comment]
OK, I think I'll wait for a reply from Giovanni. In principle, he can also cherry-pick what is needed and apply/add it to #10649.

**Comment by slava77** [comment]
for me it would still be nicer to have a separate topic branch / PR for the feature as a reference.

**Comment by ferencek** [comment]
@slava77 , in principle, I agree.

**Comment by ferencek** [comment]
OK, so I created #10760 that @gpetruc should be able to cms-merge-topic to #10649.

**Pull request merged**